open-craft/edx-analytics-pipeline
# Adds course block location to the ProblemResponse report, sourced from REST API

This change provides "course location" information and sorting for problems included in the Problem Response report.
- Course and Course Blocks data are stored to hive tables, partitioned by a formatted datetime string.
- Problem Response records are also partitioned by a formatted datetime string.
- Adds the EdxRestApiTask for contacting the REST API and caching responses.  This adds a dependency on edx-rest-api-client and its dependencies, but these modules are only required on the pipeline host machine, not the EMR instances.

**Discussions**: This change is related to [edx-analytics-data-api PR #128](https://github.com/edx/edx-analytics-data-api/pull/128), which has been discussed extensively with the Analytics team.

**Dependencies**: This is an internal pull request, intended to be merged with the [upstream PR #274](https://github.com/edx/edx-analytics-pipeline/pull/274).

**Partner information**: 3rd party-hosted open edX instance

**Testing instructions**:
1. See [edx-analytics-data-api PR 128](https://github.com/edx/edx-analytics-data-api/pull/128) for analyticstack test setup instructions.  See below for notes and alterations.
2. Before running `remote-task` to install the analytics-pipeline, create an OAuth2 client:
   
   ```
   sudo su edxapp
   ./manage.py lms --settings=devstack create_oauth2_client  \
           http://localhost:9999 \
           http://localhost:9999/complete/edx-oidc/  \
           confidential \
           --client_name "Analytics Pipeline" \
           --client_id oauth_id \
           --client_secret oauth_secret \
           --trusted
           --user=staff
   ```
3. Create a new `override.cfg` file by copying `config/devstack.cfg`. Update the parameters under `[edx-rest-api]` with your chosen OAuth2 `client_id` and `client_secret`.
   - Note that variables in `override.cfg` should not be quoted.  The quotes are taken literally, not interpreted.
4. You don't need to run the Analytics API setup, or the custom ElasticSearch.
5. You _do_ need to run the edxapp LMS service, which provides the REST API used to source the course and course block data.  You can check that the edX REST APIs are functioning by visiting:
   - [http://localhost:8000/api/courses/v1/courses/](http://localhost:8000/api/courses/v1/courses/)
   - [http://localhost:8000/api/courses/v1/blocks/](http://localhost:8000/api/courses/v1/blocks/?course_id=course-v1:edX%2BDemoX%2BDemo_Course&all_blocks=true&depth=all).  Select a course, and append `all_blocks=true&depth=all&requested_fields=children` to your querystring to see the full block hierarchy.  
6. Run the `LoadAllCourseBlocksWorkflow` task to install the pipeline from this PR's branch, and load all the courses and course blocks into your hive tables:
   
   ```
   export PIPELINE_REPO="https://github.com/open-craft/edx-analytics-pipeline.git"
   export PIPELINE_BRANCH="jill/block-location"
   export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
   export VAGRANT_PATH='..'  # or wherever your Vagrantfile is
   export REMOTE_NAME='analytics-block-location'
   remote-task --wheel-url $WHEEL_URL --vagrant-path "$VAGRANT_PATH" \
   --repo "$PIPELINE_REPO" --branch "$PIPELINE_BRANCH" \
   --remote-name "$REMOTE_NAME"  --override-config `pwd`/override.cfg  --wait \
   LoadAllCourseBlocksWorkflow --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG
   ```
7. To view the course data, query the hive table created by the task:
   
   ```
   sudo su hadoop
   hive
   hive> select * from course_list;
   hive> select * from course_blocks;
   ```
8. To see the course location field appear in the Problem Response report, run this task on the installed pipeline:
   
   ```
   sudo su hadoop
   cd /var/lib/analyticstack/analytics-block-location
   source venv/bin/activate
   cd repo
   launch-task ProblemResponseReportWorkflow --local-scheduler --n-reduce-tasks 2 --remote-log-level DEBUG --marker hdfs://localhost:9000/reports/marker`date +%s`
   ```
9. Check the configured `report_output_root` to see a report generated for each course in your LMS that has had problems answered.  The devstack.cfg file stores them under `/edx/src/problem-response-reports/`.
10. Run unit tests:
    
    ```
    vagrant ssh
    cd /var/lib/analytics-tasks/analyticstack/repo
    source ../venv/bin/activate
    export WHEEL_URL=http://edx-wheelhouse.s3-website-us-east-1.amazonaws.com/Ubuntu/precise
    export WHEEL_PYVER=python2.7
    make coverage
    ```

**Author notes and concerns**:
1. I still need to add acceptance tests.
2. Will merge this to the upstream PR branch for edX to review.

**Reviewers**
